In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sys

Проект e-learning: вариант 2
Продакт-менеджер Василий попросил вас проанализировать завершенные уроки и ответить на следующие вопросы:

1. Сколько студентов успешно сдали только один курс? (Успешная сдача — это зачёт по курсу на экзамене) (7 баллов).

2. Выяви самый сложный и самый простой экзамен: найди курсы и экзамены в рамках курса, которые обладают самой низкой и самой высокой завершаемостью*. (10 баллов)

3. По каждому предмету определи средний срок сдачи экзаменов (под сдачей понимаем последнее успешное прохождение экзамена студентом). (7 баллов) 

4. Выяви самые популярные предметы (ТОП-3) по количеству регистраций на них. А также предметы с самым большим оттоком (ТОП-3). (8 баллов)

5. Используя pandas, в период с начала 2013 по конец 2014 выяви семестр с самой низкой завершаемостью курсов и самыми долгими средними сроками сдачи курсов.  (15 баллов) 

6. Часто для качественного анализа аудитории используют подходы, основанные на сегментации. Используя python, построй адаптированные RFM-кластеры студентов, чтобы качественно оценить свою аудиторию. В адаптированной кластеризации можешь выбрать следующие метрики: R - среднее время сдачи одного экзамена, F - завершаемость курсов, M - среднее количество баллов, получаемое за экзамен. Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров. Описание подхода можно найти тут. (23 балла)

Для решения задачи проведи предварительное исследование данных и сформулируй, что должно считаться курсом. Обосновать свой выбор ты можешь с помощью фактов сдачи экзаменов, распределения студентов и уникального идентификатора курса.

*завершаемость = кол-во успешных экзаменов / кол-во всех попыток сдать экзамен

Файлы: 

assessments.csv — этот файл содержит информацию об оценках в тесте. Обычно каждый предмет в семестре включает ряд тестов с оценками, за которыми следует заключительный экзаменационный тест (экзамен).
code_module — идентификационный код предмета.

code_presentation — семестр (Идентификационный код).

id_assessment — тест (Идентификационный номер ассессмента).

assessment_type — тип теста. Существуют три типа оценивания: оценка преподавателя (TMA), компьютерная оценка (СМА), экзамен по курсу (Exam).

date — информация об окончательной дате сдачи теста. Рассчитывается как количество дней с момента начала семестра. Дата начала семестра имеет номер 0 (ноль).

weight — вес теста в % в оценке за курс. Обычно экзамены рассматриваются отдельно и имеют вес 100%; сумма всех остальных оценок составляет 100%.

courses.csv — файл содержит список предметов по семестрам.
code_module — предмет (идентификационный код).

code_presentation — семестр (идентификационный код).

module_presentation_length — продолжительность семестра в днях.

studentAssessment.csv — этот файл содержит результаты тестов студентов. Если учащийся не отправляет работу на оценку, результат не записывается в таблицу.
id_assessment — тест (идентификационный номер).

id_student — идентификационный номер студента.

date_submitted — дата сдачи теста студентом, измеряемая как количество дней с начала семестра.

is_banked — факт перезачета теста с прошлого семестра (иногда курсы перезачитывают студентам, вернувшимся из академического отпуска).

score — оценка учащегося в этом тесте. Диапазон составляет от 0 до 100. Оценка ниже 40 неудачная/неуспешная сдача теста.

studentRegistration.csv — этот файл содержит информацию о времени, когда студент зарегистрировался для прохождения курса в семестре.
code_module — предмет (идентификационный код).

code_presentation — семестр (идентификационный код)

id_student — идентификационный номер студента.

date_registration — дата регистрации студента. Это количество дней, измеренное от начала семестра (например, отрицательное значение -30 означает, что студент зарегистрировался на прохождение курса за 30 дней до его начала).

date_unregistration — дата отмены регистрации студента с предмета. У студентов, окончивших курс, это поле остается пустым.

In [2]:
assessments = pd.read_csv('assessments.csv')
courses = pd.read_csv('courses.csv')
studentAssessment = pd.read_csv('studentAssessment.csv')
studentRegistration = pd.read_csv('studentRegistration.csv')

In [3]:
courses

code_module code_presentation  module_presentation_length
0          AAA             2013J                         268
1          AAA             2014J                         269
2          BBB             2013J                         268
3          BBB             2014J                         262
4          BBB             2013B                         240
5          BBB             2014B                         234
6          CCC             2014J                         269
7          CCC             2014B                         241
8          DDD             2013J                         261
9          DDD             2014J                         262
10         DDD             2013B                         240
11         DDD             2014B                         241
12         EEE             2013J                         268
13         EEE             2014J                         269
14         EEE             2014B                         241
15         FFF             2013J                         268
16         FFF             2014J                         269
17         FFF             2013B                         240
18         FFF             2014B                         241
19         GGG             2013J                         261
20         GGG             2014J                         269
21         GGG             2014B                         241

In [4]:
assessments

code_module code_presentation  id_assessment assessment_type   date  \
0           AAA             2013J           1752             TMA   19.0   
1           AAA             2013J           1753             TMA   54.0   
2           AAA             2013J           1754             TMA  117.0   
3           AAA             2013J           1755             TMA  166.0   
4           AAA             2013J           1756             TMA  215.0   
..          ...               ...            ...             ...    ...   
201         GGG             2014J          37443             CMA  229.0   
202         GGG             2014J          37435             TMA   61.0   
203         GGG             2014J          37436             TMA  124.0   
204         GGG             2014J          37437             TMA  173.0   
205         GGG             2014J          37444            Exam  229.0   

     weight  
0      10.0  
1      20.0  
2      20.0  
3      20.0  
4      30.0  
..      ...  
201     0.0  
202     0.0  
203     0.0  
204     0.0  
205   100.0  

[206 rows x 6 columns]

# 1) Сколько студентов успешно сдали только один курс? (Успешная сдача — это зачёт по курсу на экзамене)

In [5]:
df1 = studentAssessment.merge(assessments, on='id_assessment', how='inner').drop_duplicates()

In [6]:
df1

id_assessment  id_student  date_submitted  is_banked  score  \
0                1752       11391              18          0   78.0   
1                1752       28400              22          0   70.0   
2                1752       31604              17          0   72.0   
3                1752       32885              26          0   69.0   
4                1752       38053              19          0   79.0   
...               ...         ...             ...        ...    ...   
173907          37443      527538             227          0   60.0   
173908          37443      534672             229          0  100.0   
173909          37443      546286             215          0   80.0   
173910          37443      546724             230          0  100.0   
173911          37443      558486             224          0   80.0   

       code_module code_presentation assessment_type   date  weight  
0              AAA             2013J             TMA   19.0    10.0  
1              AAA             2013J             TMA   19.0    10.0  
2              AAA             2013J             TMA   19.0    10.0  
3              AAA             2013J             TMA   19.0    10.0  
4              AAA             2013J             TMA   19.0    10.0  
...            ...               ...             ...    ...     ...  
173907         GGG             2014J             CMA  229.0     0.0  
173908         GGG             2014J             CMA  229.0     0.0  
173909         GGG             2014J             CMA  229.0     0.0  
173910         GGG             2014J             CMA  229.0     0.0  
173911         GGG             2014J             CMA  229.0     0.0  

[173912 rows x 10 columns]

In [7]:
df1.query('assessment_type == "Exam" and score >= 40')\
    .groupby('id_student',as_index=False).agg({'id_assessment':'count'})\
    .query('id_assessment == 1')\
    .id_student.count()

3802

## Успешно сдали только один курс 3802 студента

# 2) Выяви самый сложный и самый простой экзамен: найди курсы и экзамены в рамках курса, которые обладают самой низкой и самой высокой завершаемостью*

In [8]:
df2 = studentAssessment.merge(assessments, on='id_assessment', how='inner')\
    .query('assessment_type == "Exam"').drop_duplicates()

In [9]:
df2['good_exam'] = df2.score.apply(lambda x: 1 if x>=40 else 0)

In [ ]:
df2_good_exam = df2.groupby(['code_module', 'code_presentation','id_assessment'],as_index=False)\
    .agg(all_attempts = ('good_exam', 'count'),
    good_exam_sum =('good_exam', np.sum))

In [11]:
df2_good_exam['persent'] = round(df2_good_exam.good_exam_sum * 100/ df2_good_exam.all_attempts,2)

In [12]:
df2_good_exam

code_module code_presentation  id_assessment  all_attempts  good_exam_sum  \
0         CCC             2014B          24290           747            664   
1         CCC             2014J          24299          1168           1019   
2         DDD             2013B          25340           602            504   
3         DDD             2013J          25354           968            878   
4         DDD             2014B          25361           524            485   
5         DDD             2014J          25368           950            842   

   persent  
0    88.89  
1    87.24  
2    83.72  
3    90.70  
4    92.56  
5    88.63

In [13]:
df2_good_exam.sort_values('persent', ascending=False).head(1)

code_module code_presentation  id_assessment  all_attempts  good_exam_sum  \
4         DDD             2014B          25361           524            485   

   persent  
4    92.56

In [14]:
df2_good_exam.sort_values('persent', ascending=True).head(1)

code_module code_presentation  id_assessment  all_attempts  good_exam_sum  \
2         DDD             2013B          25340           602            504   

   persent  
2    83.72

## Самый сложный экзамен с ID 25340 был на курсе DDD, c 83.72 процентами удачных сдач
## Самый простой экзамен с ID 25361 был так же на курсе DDD, с 92.56 процентами удачных сдач

# 3) По каждому предмету определи средний срок сдачи экзаменов (под сдачей понимаем последнее успешное прохождение экзамена студентом). 

In [15]:
df3 = studentAssessment.merge(assessments, on='id_assessment', how='inner')

In [16]:
df3

id_assessment  id_student  date_submitted  is_banked  score  \
0                1752       11391              18          0   78.0   
1                1752       28400              22          0   70.0   
2                1752       31604              17          0   72.0   
3                1752       32885              26          0   69.0   
4                1752       38053              19          0   79.0   
...               ...         ...             ...        ...    ...   
173907          37443      527538             227          0   60.0   
173908          37443      534672             229          0  100.0   
173909          37443      546286             215          0   80.0   
173910          37443      546724             230          0  100.0   
173911          37443      558486             224          0   80.0   

       code_module code_presentation assessment_type   date  weight  
0              AAA             2013J             TMA   19.0    10.0  
1              AAA             2013J             TMA   19.0    10.0  
2              AAA             2013J             TMA   19.0    10.0  
3              AAA             2013J             TMA   19.0    10.0  
4              AAA             2013J             TMA   19.0    10.0  
...            ...               ...             ...    ...     ...  
173907         GGG             2014J             CMA  229.0     0.0  
173908         GGG             2014J             CMA  229.0     0.0  
173909         GGG             2014J             CMA  229.0     0.0  
173910         GGG             2014J             CMA  229.0     0.0  
173911         GGG             2014J             CMA  229.0     0.0  

[173912 rows x 10 columns]